### ISSUE TITLE
Unable to resolve input artifact when using global output artifact. {{workflow.outputs.artifacts.XXXX}}

### ISSUE THREAD

**Issue by soundTricker** [argoproj/argo#900]
**Is this a BUG REPORT or FEATURE REQUEST?**:
BUG REPORT

**What happened**:
When I use a global artifact outputs, Argo can't resolve the global artifact outputs variable.

**What you expected to happen**:
Argo should resolve global outputs artifacts.

**How to reproduce it (as minimally and precisely as possible)**:
Submit a code below. (This one is created from examples global-outputs.yaml)

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: global-outputs-
spec:
  entrypoint: generate-globals
  #onExit: consume-globals

  templates:
  - name: generate-globals
    steps:
    - - name: generate
        template: global-output
    - - name: consume
        template: consume-globals
        arguments:
          artifacts:
            - name: art
              from: "{{workflow.outputs.artifacts.my-global-art}}"
  - name: global-output
    container:
      image: alpine:3.7
      command: [sh, -c]
      args: ["echo -n hello world > /tmp/hello_world.txt"]
    outputs:
      parameters:
      # export a global parameter. The parameter will be programaticallyavailable in the completed
      # workflow object under: workflow.status.outputs.parameters
      - name: hello-param
        valueFrom:
          path: /tmp/hello_world.txt
        globalName: my-global-param
      # export a global artifact. The artifact will be programatically available in the completed
      # workflow object under: workflow.status.outputs.artifacts
      artifacts:
      - name: hello-art
        path: /tmp/hello_world.txt
        globalName: my-global-art

  # Once exported, global output parameters are also referenceable in later parts of the workflow.
  # In this example, the consume-globals is invoked from the onExit handler, and can reference the
  # global variable, my-global-param
  - name: consume-globals
    inputs:
      artifacts:
        - name: art
          path: /art
    container:
      image: alpine:3.7
      command: [sh, -c]
      args: ["echo {{workflow.outputs.parameters.my-global-param}}"]
```

**Anything else we need to know?**:

**Environment**:
- Argo version: 
```
argo: v2.1.1
  BuildDate: 2018-05-29T20:38:37Z
  GitCommit: ac241c95c13f08e868cd6f5ee32c9ce273e239ff
  GitTreeState: clean
  GitTag: v2.1.1
  GoVersion: go1.9.3
  Compiler: gc
  Platform: darwin/amd64
```

- Kubernetes version :
```
$ kubectl version -o yaml
clientVersion:
  buildDate: 2018-04-19T00:05:56Z
  compiler: gc
  gitCommit: dd5e1a2978fd0b97d9b78e1564398aeea7e7fe92
  gitTreeState: clean
  gitVersion: v1.9.7
  goVersion: go1.9.3
  major: "1"
  minor: "9"
  platform: darwin/amd64
serverVersion:
  buildDate: 2018-06-15T21:48:39Z
  compiler: gc
  gitCommit: eb2e43842aaa21d6f0bb65d6adf5a84bbdc62eaf
  gitTreeState: clean
  gitVersion: v1.10.4-gke.2
  goVersion: go1.9.3b4
  major: "1"
  minor: 10+
  platform: linux/amd64
```

**Other debugging information (if applicable)**:
- workflow result:
```
$ argo get <workflowname>
$ argo get  global-outputs-m68ng
Name:                global-outputs-m68ng
Namespace:           default
ServiceAccount:      default
Status:              Failed
Message:             Unable to resolve input artifact: {{workflow.outputs.artifacts.my-global-art}
Created:             Mon Jul 02 13:02:36 +0900 (1 minute ago)
Started:             Mon Jul 02 13:02:36 +0900 (1 minute ago)
Finished:            Mon Jul 02 13:02:40 +0900 (1 minute ago)
Duration:            4 seconds
Output Parameters:
  my-global-param:   hello world
Output Artifacts:
  my-global-art:     https://storage.googleapis.com/my-bucket-name-resources/global-outputs-m68ng/global-outputs-m68ng-644661169/hello-art.tgz

STEP                     PODNAME                         DURATION  MESSAGE
 ✖ global-outputs-m68ng                                            Unable to resolve input artifact: {{workflow.outputs.artifacts.my-global-art}}
 ├---✔ generate          global-outputs-m68ng-644661169  3s

```
- executor logs: There is no failed pod.
- workflow-controller logs:
```
$ kubectl logs -n kube-system $(kubectl get pods -l app=workflow-controller -n kube-system -o name)
time="2018-07-02T04:02:36Z" level=info msg="Processing workflow" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:36Z" level=info msg="Updated phase  -> Running" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:36Z" level=info msg="Steps node global-outputs-m68ng (global-outputs-m68ng) initialized Running" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:36Z" level=info msg="StepGroup node global-outputs-m68ng[0] (global-outputs-m68ng-3917597236) initialized Running" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:36Z" level=info msg="Created pod: global-outputs-m68ng[0].generate (global-outputs-m68ng-644661169)" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:36Z" level=info msg="Pod node global-outputs-m68ng[0].generate (global-outputs-m68ng-644661169) initialized Running" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:36Z" level=info msg="Workflow step group node global-outputs-m68ng[0] (global-outputs-m68ng-3917597236) not yet completed" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:36Z" level=info msg="Workflow update successful" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:37Z" level=info msg="Processing workflow" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:37Z" level=info msg="Checking for deleted pods" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:37Z" level=info msg="Workflow step group node global-outputs-m68ng[0] (global-outputs-m68ng-3917597236) not yet completed" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:38Z" level=info msg="Processing workflow" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:38Z" level=info msg="Workflow step group node global-outputs-m68ng[0] (global-outputs-m68ng-3917597236) not yet completed" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:39Z" level=info msg="Processing workflow" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:39Z" level=info msg="Workflow step group node global-outputs-m68ng[0] (global-outputs-m68ng-3917597236) not yet completed" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="Processing workflow" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="Setting node global-outputs-m68ng[0].generate (global-outputs-m68ng-644661169) outputs"
time="2018-07-02T04:02:40Z" level=info msg="Updating node global-outputs-m68ng[0].generate (global-outputs-m68ng-644661169) status Running -> Succeeded"
time="2018-07-02T04:02:40Z" level=info msg="setting workflow.outputs.parameters.my-global-param: 'hello world'" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="setting workflow.outputs.artifacts.my-global-art: {my-global-art  %!s(*int32=<nil>)  {https://storage.googleapis.com/my-bucket-name-resources/global-outputs-m68ng/global-outputs-m68ng-644661169/hello-art.tgz %!s(*v1alpha1.GitArtifact=<nil>) %!s(*v1alpha1.HTTPArtifact=<nil>) <nil> %!s(*v1alpha1.RawArtifact=<nil>)} }" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="Step group node global-outputs-m68ng[0] (global-outputs-m68ng-3917597236) successful" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="node global-outputs-m68ng[0] (global-outputs-m68ng-3917597236) phase Running -> Succeeded" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="node global-outputs-m68ng[0] (global-outputs-m68ng-3917597236) finished: 2018-07-02 04:02:40.56079676 +0000 UTC" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="overwriting workflow.outputs.parameters.my-global-param: 'hello world' -> 'hello world'" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="overwriting workflow.outputs.artifacts.my-global-art: {my-global-art  %!s(*int32=<nil>)  {https://storage.googleapis.com/my-bucket-name-resources/global-outputs-m68ng/global-outputs-m68ng-644661169/hello-art.tgz %!s(*v1alpha1.GitArtifact=<nil>) %!s(*v1alpha1.HTTPArtifact=<nil>) <nil> %!s(*v1alpha1.RawArtifact=<nil>)} }" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="StepGroup node global-outputs-m68ng[1] (global-outputs-m68ng-2910793001) initialized Running" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="SG Outbound nodes of global-outputs-m68ng-644661169 are [global-outputs-m68ng-644661169]" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="node global-outputs-m68ng[1] (global-outputs-m68ng-2910793001) phase Running -> Error" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="node global-outputs-m68ng[1] (global-outputs-m68ng-2910793001) message: Unable to resolve input artifact: {{workflow.outputs.artifacts.my-global-art}}" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="node global-outputs-m68ng[1] (global-outputs-m68ng-2910793001) finished: 2018-07-02 04:02:40.560967141 +0000 UTC" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="step group global-outputs-m68ng[1] (global-outputs-m68ng-2910793001) was unsuccessful: Unable to resolve input artifact: {{workflow.outputs.artifacts.my-global-art}}" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="Outbound nodes of global-outputs-m68ng-644661169 is [global-outputs-m68ng-644661169]" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="Outbound nodes of global-outputs-m68ng is [global-outputs-m68ng-644661169]" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="node global-outputs-m68ng (global-outputs-m68ng) phase Running -> Failed" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="node global-outputs-m68ng (global-outputs-m68ng) message: Unable to resolve input artifact: {{workflow.outputs.artifacts.my-global-art}}" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="node global-outputs-m68ng (global-outputs-m68ng) finished: 2018-07-02 04:02:40.561036295 +0000 UTC" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="Checking deamoned children of global-outputs-m68ng" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="Updated phase Running -> Failed" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="Updated message  -> Unable to resolve input artifact: {{workflow.outputs.artifacts.my-global-art}}" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="Marking workflow completed" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:40Z" level=info msg="Workflow update successful" namespace=default workflow=global-outputs-m68ng
time="2018-07-02T04:02:41Z" level=info msg="Labeled pod default/global-outputs-m68ng-644661169 completed"
time="2018-07-02T04:03:28Z" level=info msg="Alloc=3751 TotalAlloc=388813 Sys=14058 NumGC=2118 Goroutines=50"



```

**Comment by sudermanjr** [comment]
Has there been any work done to fix this?  I know a lot of work has gone toward the upcoming 2.2.2 release.

**Comment by jessesuen** [comment]
Yes, this will be in v2.2.0. Fix will be in coming in shortly.

**Comment by jessesuen** [comment]
Fixed in a5baca60d1dfb8fb8c82a936ab383d49e075cff3.
